### PR TITLE
feat: multiple compiler support

### LINKED ISSUE #39: Support MultiCompilers
### Feature Use Case

Serving a project that has both `target: 'web'` and `target: 'webworker'` e.g.

#### webpack.config.js

```
module.exports = [
  { 
    entry: 'web.js',
    target: 'web'
  },
  {
    entry: 'worker.js',
    target: 'webworker'
  }
]
```

### Feature Proposal

Documenting how this could might be possible.

### PR DESCRIPTION
**Pull request by shellscape**
This PR contains:

- [x] bugfix
- [x] feature
- [x] refactor
- [ ] tests
- [x] documentation
- [x] metadata

### Breaking Changes?

- [ ] yes
- [x] no

If yes, please describe the breakage.

### Please Describe Your Changes

This PR addresses and fixes #39 by adding proper support for MultiCompilers in both the server and client. This is mostly a refactor of the client scripts. Tests will be added at a later date, but before release, as they have now become something we can't put off any longer.

#### WIP Things

- [x] status dialog doesn't hide, or go back to success, after errors/warnings are removed
- [x] status dialog doesn't show errors/warnings

#### New Things

- **Note** this PR adds the ability for multiple compilers to perform in-place HMR in the client - something neither webpack-dev-server, nor webpack-serve are capable of.
- Progress reporting for multiple compilers in webpack is dicey, and as such, we're only going to ever show one progress indicator. In the future we'll look into how to aggregate multiple compilers' progress reporting into one progress stream.
- Only one status overlay will be displayed. Errors and warnings from multiple compilers will share the same space. This isn't ideal, but is far less complex, and the need for each compiler to have distinct sections isn't clear at the moment. We'll re-evaluate if the issue is raised.
- The browser console will display new logging similar to:
```
⬡ wps: Hot Module Replacement is active
⬡ wps: WebSocket connected (Compiler: 0) 
⬡ wps: WebSocket connected (Compiler: 1) 
```
  Where `0` and `1` are either the names of the compilers, as assigned in configuration, or the index of the respective compiler.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -72,6 +74,7 @@
     "prettier": "^1.14.3",
     "standard-version": "^4.4.0",
     "webpack": "^4.20.2",
+    "webpack-cli": "^3.1.2",
```
Comment (PlayMa256) [inline package.json]: Installed for test and forgot to remove?

Comment (shellscape) [reply package.json]: yep! killing that with 🔥 now

**Comment by shellscape** [comment]
Looks like webpack is building twice (or at least sending two `compile` hook calls) for builds that contain warnings. That's a problem for this and for webpack-nano. I'll have to resolve both before I can go back to test the status overlay.

File: .eslintrc [hunk 2]
```diff
@@ -2,7 +2,9 @@
   "extends": "shellscape",
   "globals": {
     "document": true,
+    "ʎɐɹɔosǝʌɹǝs": true,
```
Comment (jgcmarins) [inline .eslintrc]: interesting
whats is this about?

Comment (shellscape) [reply .eslintrc]: https://github.com/shellscape/webpack-plugin-serve/blob/db6f09c3412cf77a7384feafd9c24be9a19e7a34/lib/index.js#L161 😉

File: test/fixtures/multi/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  {
+    context: __dirname,
+    entry: ['./worker.js', 'webpack-plugin-serve/client'],
+    mode: 'development',
+    output: {
+      filename: './dist-worker.js',
+      path: resolve(__dirname, './output'),
+      publicPath: 'output/'
+    },
+    plugins: [serve.attach()],
```
Comment (EvanPurkhiser) [inline test/fixtures/multi/webpack.config.js:34]: Afaict this isn't documented, took me almost all day to find this, coupled with this bug I ran into here https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/289 I was pretty stuck

**Comment by EvanPurkhiser** [comment]
If anyone else comes across this PR and was stuck on how to get this working, I implemented it here: https://github.com/EvanPurkhiser/prolink-tools/commit/64b03149fe6a28020c30ea3f359c23058721c562#diff-e3030acdc5cad4eaba23b5d87e09c7384feb47b0186aabf0dd47e592e11bff62R12

The key was reusing the same instance of the plugin and calling `attach` when using it in additional compilers.

**Comment by shellscape** [comment]
Doesn't look like the Recipe was ever filled out. @PlayMa256 do you have any bandwidth to tackle that? 

For what it's work, the tests in any project are always a source of truth: https://github.com/shellscape/webpack-plugin-serve/blob/master/test/fixtures/multi/webpack.config.js

**Pull request merged**